### PR TITLE
psReadValues: Extend exceptions

### PR DESCRIPTION
**Pull request by MrWarlockX**
Extends exceptions in `psReadValues` function.

### CODE REVIEW

File: include/superior_mysqlpp/extras/prepared_statement_utils.hpp [hunk 1]
```diff
@@ -140,15 +140,19 @@ namespace SuperiorMySqlpp
...
     void psReadValues(PreparedStatementType &&ps, Args&... values)
     {
         ps.execute();
 
-        if(ps.getRowsCount() != 1 || !ps.fetch())
+        if (ps.getRowsCount() != 1 || !ps.fetch())
         {
+            if (ps.getRowsCount() == 0) {
+                throw UnexpectedZeroRowsError("psReadValues() has to have a single-row result!");
+            }
```
Comment (OpatrilPeter) [inline include/superior_mysqlpp/extras/prepared_statement_utils.hpp]: I'd prefer `else` to highlight that the throw is there for negated condition.

File: include/superior_mysqlpp/extras/prepared_statement_utils.hpp [hunk 2]
```diff
@@ -140,15 +140,19 @@ namespace SuperiorMySqlpp
      * @param ps Prepared statement object (only static statements are currently supported)
      * @param values References to variables to be loaded from query
      *               Their type must be compatible with query result types
-     * @throws std::runtime_error if more than one row is being loaded
+     * @throws UnexpectedZeroRowsError if result is empty
+     * @throws UnexpectedMultipleRowserror if more than one row is being loaded
```
Comment (OpatrilPeter) [inline include/superior_mysqlpp/extras/prepared_statement_utils.hpp]: Uppercase E in `error`.

**Review by OpatrilPeter — DISMISSED** [review]
Current version seems OK to me, as long as commits are squashed before merge.

File: include/superior_mysqlpp/extras/prepared_statement_utils.hpp [hunk 3]
```diff
@@ -140,16 +140,16 @@ namespace SuperiorMySqlpp
...
     template<typename PreparedStatementType, typename... Args, typename = typename std::enable_if<std::is_base_of<detail::StatementBase, std::remove_reference_t<PreparedStatementType>>::value>::type>
     void psReadValues(PreparedStatementType &&ps, Args&... values)
     {
         ps.execute();
 
-        if(ps.getRowsCount() != 1 || !ps.fetch())
+        if (ps.getRowsCount() != 1 || !ps.fetch())
         {
-            throw UnexpectedMultipleRowsError("psReadValues() can read only one row per query!");
+            throw UnexpectedRowCountError("psReadValues() must read exactly one row", ps.getRowsCount());
```
Comment (majcl) [inline include/superior_mysqlpp/extras/prepared_statement_utils.hpp]: I would reword it slightly. Actually psReadValuues() can still behave as it would read exactly one row. We should mention, that provided query/prepared statement read more than one line

**Review by majcl — CHANGES_REQUESTED** [review]

File: include/superior_mysqlpp/extras/prepared_statement_utils.hpp (hunk 1 above)
Comment (majcl) [reply include/superior_mysqlpp/extras/prepared_statement_utils.hpp]: I don't get it... can you explain by example?

File: include/superior_mysqlpp/exceptions.hpp [hunk 4]
```diff
@@ -108,11 +108,25 @@ namespace SuperiorMySqlpp
     };
 
     /**
-     * @brief This exception is thrown, when you attempt to read more rows in function `psReadValues`
+     * @brief Error for unexpected row count
```
Comment (majcl) [inline include/superior_mysqlpp/exceptions.hpp]: There should be more thorough explanation. You should mention, what additional information exception provides.

**Comment by majcl** [comment]
Readme.md should mention this change. With api-changes warning .

File: tests/db_access/prepared_statements.cpp [hunk 5]
```diff
@@ -603,7 +603,7 @@ go_bandit([](){
         it("can work with psReadValues (throws multiple rows error)", [&](){
             int id{};
             BlobData blob{}, binary{}, varbinary{};
-            AssertThrows(UnexpectedMultipleRowsError,
+            AssertThrows(UnexpectedRowCountError,
```
Comment (majcl) [inline tests/db_access/prepared_statements.cpp:617]: Both scenarios should be tested. 0 returned rows, and multiple returned rows.

File: packages/_debian-common/changelog [hunk 6]
```diff
@@ -1,4 +1,4 @@
-libsuperiormysqlpp (0.3.3) UNRELEASED; urgency=medium
+libsuperiormysqlpp (0.4.0) UNRELEASED; urgency=medium
```
Comment (majcl) [inline packages/_debian-common/changelog:1]: I'm wondering, whether we should mention api breaking changes within changelog. Normally these changes would force major version update, but as we are still in version 0, this is a little bit different situation. But maybe explicit API changes warning could save someone headaches.

**Review by majcl — CHANGES_REQUESTED** [review]

File: include/superior_mysqlpp/extras/prepared_statement_utils.hpp (hunk 1 above)
Comment (OpatrilPeter) [reply include/superior_mysqlpp/extras/prepared_statement_utils.hpp]: Put the other throw in else branch, which by now is already done.

File: tests/db_access/prepared_statements.cpp [hunk 7]
```diff
@@ -600,10 +600,18 @@ go_bandit([](){
             AssertThat(varbinary.getStringView(), Equals("ij\0kl"s));
         });
 
-        it("can work with psReadValues (throws multiple rows error)", [&](){
+        it("can work with psReadValues (zero rows error)", [&](){
             int id{};
             BlobData blob{}, binary{}, varbinary{};
```
Comment (smradCZ) [inline tests/db_access/prepared_statements.cpp]: I prefer made tests too much simple as possible - for instance, select varies of parameters has no added value for this test-case, the ORDERing the results in a query is useless as well.

This test-case may be reduced to form like this:
```suggestion
            int id{};
            AssertThrows(UnexpectedRowCountError,
                psReadValues("SELECT `id` FROM `test_superior_sqlpp`.`binary_data` LIMIT 0", connection, id)
            );
```

The same for `multiple rows error` test case below.

**Review by smradCZ — CHANGES_REQUESTED** [review]

File: include/superior_mysqlpp/exceptions.hpp [hunk 8]
```diff
@@ -108,11 +108,25 @@ namespace SuperiorMySqlpp
     };
 
     /**
-     * @brief This exception is thrown, when you attempt to read more rows in function `psReadValues`
+     * @brief Error for unexpected row count, accessible via `getRowCount()` method
```
Comment (majcl) [inline include/superior_mysqlpp/exceptions.hpp]: This is really strange wording. Please describe exceptions more carefully.

File: packages/_debian-common/changelog [hunk 9]
```diff
@@ -14,6 +14,10 @@ libsuperiormysqlpp (0.3.3) UNRELEASED; urgency=medium
   * Add noexcept qualifier to database driver methods
   * Fix behavior of makeHexString
 
+  [ Michal Merc ]
+  * !!! BREAKING API CHANGE !!!
```
Comment (majcl) [inline packages/_debian-common/changelog]: Maybe breaking API changes should be mentioned at the top of changelog?

**Review by majcl — CHANGES_REQUESTED** [review]

File: include/superior_mysqlpp/extras/prepared_statement_utils.hpp (hunk 1 above)
Comment (majcl) [reply include/superior_mysqlpp/extras/prepared_statement_utils.hpp]: But now there is only one exception, isn't it? There is no need for this inner `ps.getRowsCount() == 0` if

Comment (OpatrilPeter) [reply include/superior_mysqlpp/extras/prepared_statement_utils.hpp]: Yes, it's not relevant in current version anymore.

**Review by smradCZ — DISMISSED** [review]

File: tests/db_access/prepared_statements.cpp [hunk 10]
```diff
@@ -600,11 +600,17 @@ go_bandit([](){
             AssertThat(varbinary.getStringView(), Equals("ij\0kl"s));
         });
 
-        it("can work with psReadValues (throws multiple rows error)", [&](){
+        it("can work with psReadValues (zero rows error)", [&](){
```
Comment (majcl) [inline tests/db_access/prepared_statements.cpp]: Shoudln't it say something like psReadValues can detect wrong result? (here it means zero result).
Actually there should probably be `describe section` for `psReadValues`. And `"zero rows error"` should be one of its tests.

**Review by majcl — CHANGES_REQUESTED** [review]
